File: community_compute/desktop/providers.py
# -*- coding: utf-8 -*-
"""BYOK 3-provider adapter (Groq / SambaNova / NVIDIA NIM).

Adapted from universal/fleet_providers.py — but the keys come from the
VOLUNTEER's local encrypted keystore, and a job batch is SHARDED across whatever
providers the volunteer configured, translated in parallel (one thread per
provider), then any line a provider dropped is retried on another. Nothing here
ever leaves the machine except the provider API calls themselves (which the
volunteer consented to, with their own key + IP).

All three providers are OpenAI-compatible. A browser User-Agent is mandatory
(Groq sits behind Cloudflare and 403s the default Python-urllib UA — "error
code: 1010"). An optional user-configured proxy is honored (the app bundles NO
VPN; a volunteer may point at their own proxy).
"""
from __future__ import annotations

import json
import re
import ssl
import threading
import time
import urllib.error
import urllib.request

_UA = ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
       "(KHTML, like Gecko) Chrome/126.0 Safari/537.36")
_SSL = ssl.create_default_context()
_JSON = re.compile(r"\{.*\}", re.S)

# id, base url, model  (labels map to keystore ids in config.PROVIDERS)
ENDPOINTS = {
    "groq":      ("https://api.groq.com/openai/v1",      "openai/gpt-oss-120b"),
    "sambanova": ("https://api.sambanova.ai/v1",         "DeepSeek-V3.2"),
    "nim":       ("https://integrate.api.nvidia.com/v1", "meta/llama-3.1-70b-instruct"),
}

_COOLDOWN_S = 60
_COOL: dict[str, float] = {}


def _opener(proxy: str):
    if proxy:
        return urllib.request.build_opener(
            urllib.request.ProxyHandler({"http": proxy, "https": proxy}),
            urllib.request.HTTPSHandler(context=_SSL))
    return urllib.request.build_opener(urllib.request.HTTPSHandler(context=_SSL))


def _call(provider: str, key: str, sysmsg: str, usermsg: str,
          proxy: str = "", timeout: int = 120, max_tokens: int = 2500) -> str:
    base, model = ENDPOINTS[provider]
    payload = {"model": model, "temperature": 0.2, "max_tokens": max_tokens,
               "messages": [{"role": "system", "content": sysmsg},
                            {"role": "user", "content": usermsg}]}
    req = urllib.request.Request(
        base + "/chat/completions", data=json.dumps(payload).encode(), method="POST",
        headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json",
                 "User-Agent": _UA})
    body = _opener(proxy).open(req, timeout=timeout).read().decode()
    return json.loads(body)["choices"][0]["message"]["content"]


def _translate_shard(provider: str, key: str, sysmsg: str, shard: dict,
                     proxy: str, result: dict, counts: dict, lock: threading.Lock) -> None:
    if not shard:
        return
    if _COOL.get(provider, 0) > time.time():
        return
    user = ("Translate every value to Hebrew. Return ONLY a JSON object mapping the SAME "
            "ids to the Hebrew translation:\n" + json.dumps(shard, ensure_ascii=False))
    try:
        raw = _call(provider, key, sysmsg, user, proxy)
    except urllib.error.HTTPError as e:
        if e.code in (429, 402, 503):
            _COOL[provider] = time.time() + _COOLDOWN_S
        return
    except Exception:
        return
    m = _JSON.search(raw)
    if not m:
        return
    try:
        got = json.loads(m.group(0))
    except Exception:
        return
    n = 0
    with lock:
        for k in shard:
            v = got.get(k) or got.get(str(k))
            if isinstance(v, str) and v.strip():
                result[k] = v.strip()
                n += 1
        counts[provider] = counts.get(provider, 0) + n


def available(keys: dict) -> list:
    """provider ids that both have a key AND are known endpoints."""
    return [p for p in ("groq", "sambanova", "nim") if keys.get(p) and p in ENDPOINTS]


def translate_batch(keys: dict, sysmsg: str, items: dict, proxy: str = "") -> tuple[dict, dict]:
    """Shard `items` ({id:en}) across the configured providers, translate in
    parallel, retry any dropped line once on another provider. Returns
    (out {id:he}, counts {provider:lines}). Raises if NO provider is configured."""
    provs = available(keys)
    if not provs:
        raise RuntimeError("no provider keys configured")
    ids = list(items)
    result: dict = {}
    counts: dict = {}
    lock = threading.Lock()

    # round-robin shard so each provider gets a contiguous, balanced slice
    shards = {p: {} for p in provs}
    for i, k in enumerate(ids):
        shards[provs[i % len(provs)]][k] = items[k]

    threads = [threading.Thread(target=_translate_shard,
                                args=(p, keys[p], sysmsg, shards[p], proxy, result, counts, lock))
               for p in provs]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    # retry whatever any provider dropped, on a NOT-cooling provider
    missing = {k: items[k] for k in ids if k not in result}
    if missing:
        for p in provs:
            if _COOL.get(p, 0) > time.time():
                continue
            _translate_shard(p, keys[p], sysmsg, missing, proxy, result, counts, lock)
            missing = {k: items[k] for k in ids if k not in result}
            if not missing:
                break
    return result, counts

File: community_compute/android/lib/providers.dart
// BYOK 3-provider adapter — shard a batch across the volunteer's providers, run
// them in parallel, retry any dropped line. Mirrors desktop/providers.py.
// Browser User-Agent is mandatory (Groq is behind Cloudflare). An optional
// user proxy is honored (the app bundles NO VPN).
//
// Each provider's slice is translated in SMALL CHUNKS (~15 lines/call) so no
// single request is huge — this matters when only ONE (possibly slow) provider
// is configured. Every call reports a per-provider STATUS so the UI can say WHY
// it failed: ok | 401 | 403 | 429 | 402 | 503 | http | timeout | neterr | parse.
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:http/io_client.dart';
import 'config.dart';

const _ua = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
    '(KHTML, like Gecko) Chrome/126.0 Safari/537.36';
const _chunk = 10; // max lines per provider call (smaller = friendlier to slow free tiers)

final _cooldown = <String, DateTime>{};

Map<String, List<String>> get _ep {
  final m = <String, List<String>>{};
  for (final p in Config.providers) {
    m[p[1]] = [p[2], p[3]]; // id -> [base, model]
  }
  return m;
}

http.Client _client(String proxy) {
  if (proxy.trim().isEmpty) return http.Client();
  final hc = HttpClient();
  final u = Uri.parse(proxy);
  hc.findProxy = (uri) => 'PROXY ${u.host}:${u.port}';
  return IOClient(hc);
}

List<String> available(Map<String, String> keys) =>
    ['groq', 'sambanova', 'nim'].where((p) => (keys[p] ?? '').trim().isNotEmpty).toList();

/// One HTTP call for a small chunk. Returns (translated, status).
Future<(Map<String, String>, String)> _call(String provider, String key, String sys,
    Map<String, dynamic> chunk, String proxy) async {
  if (chunk.isEmpty) return (<String, String>{}, 'ok');
  final cd = _cooldown[provider];
  if (cd != null && cd.isAfter(DateTime.now())) return (<String, String>{}, 'cooldown');
  final ep = _ep[provider]!;
  final user = 'Translate every value to Hebrew. Return ONLY a JSON object mapping the SAME '
      'ids to the Hebrew translation:\n${jsonEncode(chunk)}';
  final client = _client(proxy);
  try {
    final r = await client.post(Uri.parse('${ep[0]}/chat/completions'),
        headers: {
          'Authorization': 'Bearer $key',
          'Content-Type': 'application/json',
          'User-Agent': _ua,
        },
        body: jsonEncode({
          'model': ep[1], 'temperature': 0.2, 'max_tokens': 2000, 'stream': false,
          'messages': [
            {'role': 'system', 'content': sys},
            {'role': 'user', 'content': user},
          ],
        })).timeout(const Duration(seconds: 120));
    if (r.statusCode == 429 || r.statusCode == 402 || r.statusCode == 503) {
      _cooldown[provider] = DateTime.now().add(const Duration(seconds: 60));
      return (<String, String>{}, '${r.statusCode}');
    }
    if (r.statusCode == 401 || r.statusCode == 403) {
      _cooldown[provider] = DateTime.now().add(const Duration(seconds: 120)); // bad key — don't hammer
      return (<String, String>{}, '${r.statusCode}');
    }
    if (r.statusCode >= 400) {
      _cooldown[provider] = DateTime.now().add(const Duration(seconds: 30));
      return (<String, String>{}, 'http');
    }
    String content;
    try {
      content = jsonDecode(utf8.decode(r.bodyBytes))['choices'][0]['message']['content'] as String;
    } catch (_) {
      return (<String, String>{}, 'parse');
    }
    final m = RegExp(r'\{.*\}', dotAll: true).firstMatch(content);
    if (m == null) return (<String, String>{}, 'parse');
    Map got;
    try { got = jsonDecode(m.group(0)!) as Map; } catch (_) { return (<String, String>{}, 'parse'); }
    final out = <String, String>{};
    chunk.forEach((k, _) {
      final v = got[k] ?? got[k.toString()];
      if (v is String && v.trim().isNotEmpty) out[k] = v.trim();
    });
    return (out, out.isEmpty ? 'parse' : 'ok');
  } on TimeoutException {
    _cooldown[provider] = DateTime.now().add(const Duration(seconds: 30)); // slow key — back off briefly
    return (<String, String>{}, 'timeout');
  } on SocketException {
    _cooldown[provider] = DateTime.now().add(const Duration(seconds: 20));
    return (<String, String>{}, 'neterr');
  } on HttpException {
    _cooldown[provider] = DateTime.now().add(const Duration(seconds: 20));
    return (<String, String>{}, 'neterr');
  } catch (_) {
    _cooldown[provider] = DateTime.now().add(const Duration(seconds: 20));
    return (<String, String>{}, 'neterr');
  } finally {
    client.close();
  }
}

/// Translate a provider's whole slice in chunks of [_chunk]. Stops early if the
/// provider goes on cooldown (429/402/503). Returns (translated, worst status).
Future<(Map<String, String>, String)> _shard(String provider, String key, String sys,
    Map<String, dynamic> shard, String proxy) async {
  if (shard.isEmpty) return (<String, String>{}, 'ok');
  // a HARD failure means the whole key/provider is unhealthy right now — stop
  // spending another 120s per chunk on it (a bad/slow key used to hang the batch
  // for up to ~15 min, showing "מתרגם" while producing nothing). 'parse' is
  // per-chunk (the model may fix the next one), so it does NOT break.
  const hard = {'timeout', 'neterr', '401', '403', 'http', '429', '402', '503', 'cooldown'};
  final ids = shard.keys.toList();
  final out = <String, String>{};
  var worst = 'ok';
  for (var i = 0; i < ids.length; i += _chunk) {
    final sub = {for (final k in ids.skip(i).take(_chunk)) k: shard[k]};
    final (got, stat) = await _call(provider, key, sys, sub, proxy);
    out.addAll(got);
    if (stat != 'ok' && got.isEmpty) {
      worst = stat;
      if (hard.contains(stat)) break;   // fail fast — don't grind all chunks on a dead key
    }
  }
  return (out, out.isEmpty ? worst : 'ok');
}

/// Returns (out {id:he}, counts {provider:lines}, statuses {provider:code}).
Future<(Map<String, String>, Map<String, int>, Map<String, String>)> translateBatch(
    Map<String, String> keys, String sys, Map<String, dynamic> items, String proxy) async {
  final provs = available(keys);
  if (provs.isEmpty) throw StateError('no provider keys');
  final ids = items.keys.toList();
  final shards = {for (final p in provs) p: <String, dynamic>{}};
  for (var i = 0; i < ids.length; i++) {
    shards[provs[i % provs.length]]![ids[i]] = items[ids[i]];
  }
  final results = await Future.wait(
      provs.map((p) => _shard(p, keys[p]!, sys, shards[p]!, proxy)));
  final out = <String, String>{};
  final counts = <String, int>{};
  final statuses = <String, String>{};
  for (var i = 0; i < provs.length; i++) {
    out.addAll(results[i].$1);
    counts[provs[i]] = (counts[provs[i]] ?? 0) + results[i].$1.length;
    statuses[provs[i]] = results[i].$2;
  }
  // retry dropped lines on a not-cooling provider
  var missing = {for (final k in ids) if (!out.containsKey(k)) k: items[k]};
  if (missing.isNotEmpty) {
    for (final p in provs) {
      final cd = _cooldown[p];
      if (cd != null && cd.isAfter(DateTime.now())) continue;
      final (got, stat) = await _shard(p, keys[p]!, sys, missing, proxy);
      out.addAll(got);
      counts[p] = (counts[p] ?? 0) + got.length;
      if (got.isNotEmpty) statuses[p] = 'ok';
      else if (statuses[p] == 'ok') statuses[p] = stat;
      missing = {for (final k in ids) if (!out.containsKey(k)) k: items[k]};
      if (missing.isEmpty) break;
    }
  }
  return (out, counts, statuses);
}
